feat: PreDelete hooks support (Issue #13975)

pedro-cs-ribeiro opened the pull request:
Closes #13975


This PR introduces PreDelete hooks for Argo CD. PreDelete hooks allow users to execute tasks before an application's resources are deleted, providing a way to perform graceful shutdowns, backup operations, or other cleanup tasks that should happen before resource deletion.

Implementation details:

The implementation follows the same [pattern established for PostDelete hooks](https://github.com/argoproj/argo-cd/pull/16595):

- Adds PreDeleteFinalizerName constant
- Implements Pre/UnSet/HasPreDeleteFinalizer methods
- Adds hook detection logic via isPreDeleteHook()
- Implements hook execution methods via executePreDeleteHooks() and cleanupPreDeleteHooks()
- Updates application controller to respect the proper deletion order

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [X] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [X] Does this PR require documentation updates?
* [X] I've updated documentation as required by this PR.
* [X] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [X] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [X] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [X] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).


https://github.com/user-attachments/assets/7ea01afb-a13f-475e-b865-d76cabb39060

andrii-korotkov-verkada commented on controller/hook.go:
```diff
@@ -41,7 +46,151 @@ func isPostDeleteHook(obj *unstructured.Unstructured) bool {
+			}
+		}
+		if hookHealth.Status == health.HealthStatusProgressing {
+			progressingHooksCnt++
```
Suggest to use full names, i.e. `Count`

pedro-cs-ribeiro replied:
Refactored to `progressingHooksCnt`

andrii-korotkov-verkada replied:
I meant `progressingHooksCount`

andrii-korotkov-verkada commented on controller/hook.go:
```diff
@@ -41,7 +46,151 @@ func isPostDeleteHook(obj *unstructured.Unstructured) bool {
+			hookHealth = &health.HealthStatus{
+				Status: health.HealthStatusHealthy,
+			}
+		}
```
Is this necessary? I don't know if assuming healthy for when there's no health would not have any undesired side effects.

pedro-cs-ribeiro replied:
For pre-delete hooks, defaulting to "Healthy" when no health check is defined prevents the process from getting stuck waiting for an undefined condition. 
This approach is consistent with how we handle other hook types on the code.

andrii-korotkov-verkada commented on server/application/deepinformer_test.go:
```diff
@@ -86,7 +85,7 @@ func Test_deepCopyAppProjectClient_List(t *testing.T) {
 			AppProjectInterface: func() clientset.AppProjectInterface {
 				appProject := mocks.AppProjectInterface{}
-				appProject.On("List", context.Background(), metav1.ListOptions{}).Return(nil, errors.New("error"))
+				appProject.On("List", t.Context(), metav1.ListOptions{}).Return(nil, errors.New("error"))
```
Need to rebase, this was fixed in another PR

pedro-cs-ribeiro replied:
Rebased to latest upstream master

andrii-korotkov-verkada replied:
Looks like rebase wasn't fully done or wasn't to the latest master, since these changes are still here.

andrii-korotkov-verkada commented on controller/appcontroller.go:
```diff
@@ -1193,26 +1193,53 @@ func (ctrl *ApplicationController) finalizeApplicationDeletion(app *appv1.Applic
 	destCluster, err := argo.GetDestinationCluster(context.Background(), app.Spec.Destination, ctrl.db)
 	if err != nil {
-		logCtx.Warnf("Unable to get destination cluster: %v", err)
+		logCtx.Warnf("Unable to locate cluster URL for Application being deleted: %v", err)
```
Is this accurate? What if we can locate a URL, but there's some other error getting the cluster?

andrii-korotkov-verkada commented on controller/appcontroller.go:
```diff
@@ -1193,26 +1193,53 @@ func (ctrl *ApplicationController) finalizeApplicationDeletion(app *appv1.Applic
 		app.UnSetPostDeleteFinalizer()
+		if app.HasPreDeleteFinalizer() {
+			app.UnSetPreDeleteFinalizer()
+		}
```
Why doesn't `app.UnSetPostDeleteFinalizer()` require such a check?

andrii-korotkov-verkada commented on controller/hook.go:
```diff
@@ -14,19 +14,24 @@ import (
 
 func isHook(obj *unstructured.Unstructured) bool {
-	return hook.IsHook(obj) || isPostDeleteHook(obj)
+	return hook.IsHook(obj) || isPostDeleteHook(obj) || isPreDeleteHook(obj)
```
Why are pre- and post- delete hooks are handled separately?

pedro-cs-ribeiro replied:
I refactored the isHook function to support all Hook and future hook types: [`ba01602` (#22288)](https://github.com/argoproj/argo-cd/pull/22288/commits/ba01602ab9ffe1e527d2945b017cc003f77f59dd)

andrii-korotkov-verkada commented on controller/hook.go:
```diff
@@ -41,7 +46,151 @@ func isPostDeleteHook(obj *unstructured.Unstructured) bool {
+	return true, nil
+}
+
+func (ctrl *ApplicationController) executePostDeleteHooks(app *appv1.Application, proj *appv1.AppProject, liveObjs map[kube.ResourceKey]*unstructured.Unstructured, config *rest.Config, logCtx *log.Entry) (bool, error) {
```
Can we unify some code with post-detele hooks?

pedro-cs-ribeiro replied:
I refactored the execute and cleanup functions, creating generic functions to eleminate the duplicated code: [`92dcf90` (#22288)](https://github.com/argoproj/argo-cd/pull/22288/commits/92dcf90770ae64b791dcabcb46d382b5b9bc4ae4)

andrii-korotkov-verkada commented on ui/src/app/settings/components/repos-list/repos-list.tsx:
```diff
@@ -208,31 +208,34 @@ export class ReposList extends React.Component<
-                        (httpsValues.type != 'git' && 'Bearer token is only supported for Git BitBucket Data Center repositories'),
-                    tlsClientCertKey: !httpsValues.tlsClientCertKey && httpsValues.tlsClientCertData && 'TLS client cert key is required if TLS client cert is given.'
+                        (validURLValues.password && validURLValues.bearerToken && 'Either the password or the bearer token must be set, but not both.') ||
+                        (validURLValues.type != 'git' && 'Bearer token is only supported for Git BitBucket Data Center repositories.')
```
Why are these changes needed?

andrii-korotkov-verkada commented on ui/src/app/settings/components/repos-list/repos-list.tsx:
```diff
@@ -860,9 +863,9 @@ export class ReposList extends React.Component<
-    private isHTTPSUrl(url: string) {
-        if (url.match(/^https:\/\/.*$/gi)) {
+    // Whether url is a http or https url
+    private isHTTPOrHTTPSUrl(url: string) {
```
Can we separate these changes into a different PR, please?

xrow commented:
Yesterday I was implementing a workaround because of this missing feature. Now I think holy sh*t cow there it is.

pedro-cs-ribeiro commented:
@andrii-korotkov-verkada Thanks for the Review! I rebased the code correctly now and implemented some changes according to your review, please check.

leoluz commented on docs/user-guide/resource_hooks.md:
```diff
@@ -9,6 +9,7 @@ and after a Sync operation. Hooks can also be run if a Sync operation fails at a
 Kubernetes rolling update strategy.
 * Using a `PostSync` hook to run integration and health checks after a deployment.
 * Using a `SyncFail` hook to run clean-up or finalizer logic if a Sync operation fails.
+* Using a `PreDelete` hook to run tasks before an Application's resources are deleted. This can be useful for graceful shutdowns, backup operations, or other pre-deletion tasks. Similar to `PostDelete` hooks, `PreDelete` hooks are executed when the application is being deleted, and are managed separately from the application lifecycle.
```
What happens if I remove a resource from git and sync with prune option enabled? Will the PreDelete hook be executed? It is better to clarify in the docs in which exact conditions the hook gets triggered.

pedro-cs-ribeiro replied:
Agreed. I updated the docs [`dcc55ec` (#22288)](https://github.com/argoproj/argo-cd/pull/22288/commits/dcc55ec62236846f6bfff0eeac69aa384af7bb2d) to clarify the conditions where the hook gets triggered

leoluz commented on controller/hook.go:
We should have unit-tests validating this logic

pedro-cs-ribeiro replied:
Added the ``hook_test.go`` tests here: [`33c2007` (#22288)](https://github.com/argoproj/argo-cd/pull/22288/commits/33c200730cc18f6ff3cce3610a19b4ab5cea8a3d)

leoluz commented on controller/appcontroller.go:
```diff
@@ -1193,26 +1193,51 @@ func (ctrl *ApplicationController) finalizeApplicationDeletion(app *appv1.Applic
+	if app.HasPreDeleteFinalizer() {
+		objsMap, err := ctrl.getPermittedAppLiveObjects(destCluster, app, proj, projectClusters)
+		if err != nil {
+			return err
```
Please return the context alongside with errors. 
Example:
```golang
return fmt.Errorf("error getting permitted app live objects: %w", err)
```

pedro-cs-ribeiro replied:
Updated the code to return the context now with the error: [`6a864c8` (#22288)](https://github.com/argoproj/argo-cd/pull/22288/commits/6a864c813c8b03259070fd0128debfc2e0291de8)

leoluz commented on controller/appcontroller.go:
```diff
@@ -1193,26 +1193,51 @@ func (ctrl *ApplicationController) finalizeApplicationDeletion(app *appv1.Applic
+
+		done, err := ctrl.executePreDeleteHooks(app, proj, objsMap, config, logCtx)
+		if err != nil {
+			return err
```
Please return the context alongside with errors. 
Example:
```golang
return fmt.Errorf("error getting permitted app live objects: %w", err)
```

pedro-cs-ribeiro replied:
Updated the code to return the context now with the error: [`6a864c8` (#22288)](https://github.com/argoproj/argo-cd/pull/22288/commits/6a864c813c8b03259070fd0128debfc2e0291de8)

leoluz commented on controller/appcontroller.go:
```diff
@@ -1193,26 +1193,51 @@ func (ctrl *ApplicationController) finalizeApplicationDeletion(app *appv1.Applic
+		// PreDelete hooks are done - remove the finalizer so we can continue with deletion
+		app.UnSetPreDeleteFinalizer()
+		if err := ctrl.updateFinalizers(app); err != nil {
+			return err
```
Please return the context alongside with errors. 
Example:
```golang
return fmt.Errorf("error getting permitted app live objects: %w", err)
```

pedro-cs-ribeiro replied:
Updated the code to return the context now with the error: [`6a864c8` (#22288)](https://github.com/argoproj/argo-cd/pull/22288/commits/6a864c813c8b03259070fd0128debfc2e0291de8)

leoluz commented on controller/appcontroller.go:
```diff
@@ -1289,6 +1314,23 @@ func (ctrl *ApplicationController) finalizeApplicationDeletion(app *appv1.Applic
+	if app.HasPreDeleteFinalizer("cleanup") {
+		objsMap, err := ctrl.getPermittedAppLiveObjects(destCluster, app, proj, projectClusters)
+		if err != nil {
+			return err
```
Please return the context alongside with errors. 
Example:
```golang
return fmt.Errorf("error getting permitted app live objects: %w", err)
```

pedro-cs-ribeiro replied:
Updated the code to return the context now with the error: [`6a864c8` (#22288)](https://github.com/argoproj/argo-cd/pull/22288/commits/6a864c813c8b03259070fd0128debfc2e0291de8)

leoluz commented on controller/appcontroller.go:
```diff
@@ -1289,6 +1314,23 @@ func (ctrl *ApplicationController) finalizeApplicationDeletion(app *appv1.Applic
+
+		done, err := ctrl.cleanupPreDeleteHooks(objsMap, config, logCtx)
+		if err != nil {
+			return err
```
Please return the context alongside with errors. 
Example:
```golang
return fmt.Errorf("error getting permitted app live objects: %w", err)
```

pedro-cs-ribeiro replied:
Updated the code to return the context now with the error: [`6a864c8` (#22288)](https://github.com/argoproj/argo-cd/pull/22288/commits/6a864c813c8b03259070fd0128debfc2e0291de8)

leoluz reviewed (changes requested):
Thank you for the contribution!
Please check my comments.

pedro-cs-ribeiro commented:
Thanks for the review @leoluz ! I did some changes regarding your comments, PTAL

tony-vsx commented:
Any update?

dfry commented:
any update?

joek-usmobile commented:
👍

mark-anaplan commented:
It would be really great if this was merged! Is there anything still needing doing to get it over the line?

mark-anaplan commented:
@pedro-cs-ribeiro Many thanks, very much appreciated 👍

tony24681379 commented:
@agaudreault @crenshaw-dev could you help for this feature?

mark-anaplan commented:
@leoluz I think that your review comments have been addressed - any chance you could take another quick look please?

davidswimbird commented:
Status? Just wondering if I should wait for this to be released - or hack my way around it.

agaudreault reviewed (changes requested):
Overall LGTM.

pedro-cs-ribeiro commented on controller/appcontroller_test.go:
```diff
@@ -409,6 +409,38 @@ metadata:
+    },
+    "annotations": {
+      "argocd.argoproj.io/hook": "PreDelete",
+      "argocd.argoproj.io/hook-delete-policy": "HookSucceeded"
```
Addressed in https://github.com/argoproj/argo-cd/pull/22288/commits/0d35eed65eaf3ce29e65a9f241074be20200ae9e

pedro-cs-ribeiro commented:
Thanks for the review @agaudreault !

I just addressed all your comments and merged with the latest code, please take another look whenever you can.

agaudreault reviewed (approved):
LGTM. I think the current code is solid and a good foundation for the future. User experience can be adapted as the feature usage grows 👍
